projects/gigya/src/auth/store/effects/gigya-user-token.effect.ts

```diff
@@ -0,0 +1,112 @@
+import { Injectable } from '@angular/core';
+import { Actions, Effect, ofType } from '@ngrx/effects';
+import { Observable, of } from 'rxjs';
+import { catchError, map, mergeMap } from 'rxjs/operators';
+import {
+  UserToken,
+  OCC_USER_ID_CURRENT,
+  ErrorModel,
+  HttpErrorModel,
+  GlobalMessageService,
+  GlobalMessageType,
+} from '@spartacus/core';
+import { GigyaUserAuthenticationTokenService } from '../../services/user-authentication/gigya-user-authentication-token.service';
+import { GigyaAuthActions } from '../actions';
+import { HttpErrorResponse } from '@angular/common/http';
+
+export const UNKNOWN_ERROR = {
+  error: 'unknown error',
+};
+
+const circularReplacer = () => {
+  const seen = new WeakSet();
+  return (_key: any, value: any) => {
+    if (typeof value === 'object' && value !== null) {
+      if (seen.has(value)) {
+        return;
+      }
+      seen.add(value);
+    }
+    return value;
+  };
+};
+
+@Injectable()
+export class GigyaUserTokenEffects {
+  @Effect()
+  loadGigyaUserToken$: Observable<
+    GigyaAuthActions.GigyaUserTokenAction
+  > = this.actions$.pipe(
+    ofType(GigyaAuthActions.LOAD_GIGYA_USER_TOKEN),
+    map((action: GigyaAuthActions.LoadGigyaUserToken) => action.payload),
+    mergeMap(({ UID, UIDSignature, signatureTimestamp, idToken, baseSite }) =>
+      this.userTokenService
+        .loadTokenUsingCustomFlow(
+          UID,
+          UIDSignature,
+          signatureTimestamp,
+          idToken,
+          baseSite
+        )
+        .pipe(
+          map((token: UserToken) => {
+            const date = new Date();
+            date.setSeconds(date.getSeconds() + token.expires_in);
+            token.expiration_time = date.toJSON();
+            token.userId = OCC_USER_ID_CURRENT;
+            return new GigyaAuthActions.LoadUserTokenSuccess(token);
```

Comment:
Could you also pass to success action the whole payload from the action initializing this effect.
Eg. `{ initActionPayload: action.payload, token: token }`

Similar in error action. Adding retries/error handling is much easier then.